src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -377,91 +349,66 @@ private float avgRowSizeInBytes()
     }
 
     /**
-     * Build a {@link RangeIterator.Builder} from the given list of expressions by applying given operation (OR/AND).
+     * Build a {@link Plan} from the given list of expressions by applying given operation (OR/AND).
      * Building of such builder involves index search, results of which are persisted in the internal resources list
      *
-     * @param op The operation type to coalesce expressions with.
-     * @param expressions The expressions to build range iterator from (expressions with not results are ignored).
-     *
-     * @return range iterator builder based on given expressions and operation type.
+     * @param builder The plan node builder which receives the built index scans
+     * @param expressions The expressions to build the plan from
      */
-    public RangeIterator buildRangeIteratorForExpressions(Operation.OperationType op, Collection<Expression> expressions)
+    public void buildPlanForExpressions(Plan.Builder builder, Collection<Expression> expressions)
     {
-        assert !expressions.isEmpty() : "expressions should not be empty for " + op + " in " + filterOperation;
+        Operation.OperationType op = builder.type;
+        assert !expressions.isEmpty() : "expressions should not be empty for " + op + " in " + command.rowFilter().root();
 
         // VSTODO move ANN out of expressions and into its own abstraction? That will help get generic ORDER BY support
         Collection<Expression> exp = expressions.stream().filter(e -> e.operation != Expression.Op.ANN).collect(Collectors.toList());
-        boolean defer = op == Operation.OperationType.OR || RangeIntersectionIterator.shouldDefer(exp.size());
-        RangeIterator.Builder builder = op == Operation.OperationType.OR
-                                        ? RangeUnionIterator.builder()
-                                        : RangeIntersectionIterator.builder(RangeIntersectionIterator.INTERSECTION_CLAUSE_LIMIT);
+        boolean defer = builder.type == Operation.OperationType.OR || RangeIntersectionIterator.shouldDefer(exp.size());
 
         Set<Map.Entry<Expression, NavigableSet<SSTableIndex>>> view = referenceAndGetView(op, exp).entrySet();
-
+        List<RangeIterator> indexIterators = new ArrayList<>();
         try
         {
             for (Map.Entry<Expression, NavigableSet<SSTableIndex>> e : view)
             {
-                @SuppressWarnings("resource") // RangeIterators are closed by releaseIndexes
-                RangeIterator index = TermIterator.build(e.getKey(), e.getValue(), mergeRange, queryContext, defer, Integer.MAX_VALUE);
-
-                builder.add(index);
+                Expression predicate = e.getKey();
+                RangeIterator iterator = TermIterator.build(predicate, e.getValue(), mergeRange, queryContext, defer, Integer.MAX_VALUE);
+                indexIterators.add(iterator);
+                keyIterators.put(predicate, iterator);
```

Comment:
> If that is the only usage, would it be possible/simpler to store a reference to the iterator in the IndexScan object created below, and then when we need to closeUnreferencedIterators, we just call close on the IndexScan objects that don't make it into the final execution and the IndexScan#execute method just returns the RangeIterator?

You won't believe, but it was almost like that in one of the intermediate stages I have (however a node would take a factory for an iterator, not the iterator). While technically it would be simpler, and we could even test it (you can use LongIterator) I'm not super happy with the Plan depending on the iterators when we are using the plan to control building the final iterator. So that would be a dependency cycle. 

The fact that we need iterators to construct the plan at the moment is very accidental and temporary - we need them only because we have no other way to get good cardinality estimates. So we get ahead and make those iterators earlier and cache them so when the Plan needs to construct them for execution, we don't repeat the work. 
To me it looks a lot cleaner. 

>  If that is the case, what do you think about calling keyIterators.remove(predicate) in getKeysFromIndex to ensure we never attempt to re-run the same predicate?

Yes, that looks like a great idea to me!

Reply:
Even better, I think if we remove the iterators from the map in `getKeysFromIndex`, the logic for cleaning up the unused ones gets a lot simpler! We can just close the ones that are left in the map.

BTW: I'm changing the map to a multi map because we have no guarantee expressions are unique.

Reply:
Done.

Reply:
> The fact that we need iterators to construct the plan at the moment is very accidental and temporary

This helps me understand the design, and I support trying to find ways to remove the dependency. It seems like range queries require quite a bit of up front work to build all of the individual postings list readers, so it would be good to remove it, if possible. I support keeping the RangeIterators separate.